compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2832,20 +2833,33 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     ));
                 }
             }
-            ObligationCauseCode::RepeatElementCopy { is_const_fn } => {
+            ObligationCauseCode::RepeatElementCopy { is_constable } => {
                 err.note(
                     "the `Copy` trait is required because this value will be copied for each element of the array",
                 );
 
-                if is_const_fn {
-                    err.help(
-                        "consider creating a new `const` item and initializing it with the result \
-                        of the function call to be used in the repeat position, like \
-                        `const VAL: Type = const_fn();` and `let x = [VAL; 42];`",
-                    );
+                let value_kind = match is_constable {
+                    IsConstable::Fn => Some("the result of the function call"),
+                    IsConstable::Ctor => Some("the result of the constructor"),
+                    _ => None
+                };
+                if let Some(value_kind) = value_kind &&
+                    let span = err.span.primary_spans()[0] &&
+                    let sm = tcx.sess.source_map() &&
+                    let Ok(snip) = sm.span_to_snippet(span)
+                {
+                    let line_start = sm.span_extend_to_line(span).shrink_to_lo();
+                    let indentation = sm.indentation_before(span).unwrap_or_default();
+                    let help_msg = format!(
+                        "consider creating a new `const` item and initializing it with {value_kind} \
+                        to be used in the repeat position");
+                    err.multipart_suggestion(help_msg, vec![
+                        (line_start, format!("{indentation}const VALUE: /* .. */ = {snip};\n")),
+                        (span, "VALUE".to_string())
+                    ], Applicability::HasPlaceholders);
```

Comment:
...then you can do something like this
```suggestion
                let sm = tcx.sess.source_map();
                if let Some(value_kind) = value_kind
                    && let Ok(snip) = sm.span_to_snippet(repeat_span)
                {
                    let indentation = sm.indentation_before(sugg_span).unwrap_or_default();
                    let help_msg = format!(
                        "consider creating a new `const` item and initializing it with {value_kind} \
                         to be used in the repeat position");
                    err.multipart_suggestion(
                        help_msg,
                        vec![
                            (sugg_span.shrink_to_lo(), format!("const VALUE: /* .. */ = {snip};\n{indentation}")),
                            (repeat_span, "VALUE".to_string())
                        ],
                        Applicability::HasPlaceholders,
                    );
```